### PR TITLE
Prevent reading or writing to a closed cache file channel

### PR DESCRIPTION
**Pull request by tlrx**
This pull request reintroduces the refcounting around the cache file channel (ie, `SharedBytes`) so that it cannot be closed while read or write operations are queued or being executed.

Thanks to `SharedBytes` already implementing `RefCounted`, we can increment the refcount of the cache file channel before populating the cache, and decrement the refcount back when all required gaps are filled and read. 

It also completes the gaps before releasing the `CacheFileRegion` and `SharedBytes`, therefore it supersedes #107359.

Relates ES-8531

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java [hunk 1]
```diff
@@ -298,8 +300,31 @@ private IO(final int sharedBytesPos, MappedByteBuffer mappedByteBuffer) {
             this.mappedByteBuffer = mappedByteBuffer;
         }
 
+        @Override
+        public boolean tryIncRef() {
+            return SharedBytes.this.tryIncRef();
```
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java:305]: I think this means all reads and writes will now have one ref-count that all threads contend on. This seems sub optimal relative to having something more optimistic?

**Review by henningandersen — COMMENTED** [review]
Left a few comments. I wonder about the motivation for the PR, I see the linked jira, but another approach could be to ensure any such merges are stopped/aborted before we close the shared bytes? I am not entirely sure how this PR helps, except that it could result in direct blob store accesses instead of cache use?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 2]
```diff
@@ -878,13 +884,17 @@ private static void throwAlreadyEvicted() {
          */
         boolean tryRead(ByteBuffer buf, long offset) throws IOException {
             SharedBytes.IO ioRef = this.io;
-            if (ioRef != null) {
-                int readBytes = ioRef.read(buf, getRegionRelativePosition(offset));
-                if (isEvicted()) {
-                    buf.position(buf.position() - readBytes);
-                    return false;
+            if (ioRef != null && ioRef.tryIncRef()) {
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:887]: Yes, here we can catch the exception.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java:305]: Agree it is suboptimal but I haven't found a better way to protect read/write access.

**Comment by tlrx** [comment]
> I wonder about the motivation for the PR, I see the linked jira, but another approach could be to ensure any such merges are stopped/aborted before we close the shared bytes?

I wonder if that will be sufficient? I was under the assumption that various things (like warming) could access Lucene files while the shared bytes is closing/closed. By reintroducing some refcounting like this I wanted to make this simpler to reason about, and just rely on tasks to be processed to close the underlying file.

> I am not entirely sure how this PR helps, except that it could result in direct blob store accesses instead of cache use?

I think the current behavior would fail or corrupt the engine, which looked worst to me (but maybe not that bad on serverless since the node is shutting down anyway)

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java]: I pushed ab1351f215341254f091b26376aa4fd6288841c8 to catch the ClosedChannelExc.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java]: I reduced a bit the contention by having each IO using its own refcount in ab1351f215341254f091b26376aa4fd6288841c8

**Comment by tlrx** [comment]
@henningandersen and I talked about this change and agreed that keeping the SharedBytes around until read/write tasks are processed is not enough: the tasks require more services to complete (object store being the most obvious one) and they are not guaranteed to be operable after the plugin is closed.

I'm going to close this PR for now and reopen #107359 which contains bits I'd like to merge. I'm also going to keep a related private pull request since it adds a test for a case we're interested to fix.

**Pull request closed without merging**